### PR TITLE
Allow almost all printable ASCII characters in environment variables

### LINKED ISSUE #53201: Environment variables containing colon not allowed
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind feature

**Problem:**
Kubernetes does not support `:` in environment variables. I can see that the subject has recently been discussed in [PR #48986](https://github.com/kubernetes/kubernetes/pull/48986). .NET Core uses `:` to replace config variables as mentioned in [this comment](https://github.com/kubernetes/kubernetes/pull/48986#issuecomment-318628326). The limitation in kubernetes requires me to perform some ugly hacks to override the config values.

**Proposed Solution:**
I'm in favor of the sentiment that kubernetes should not censor environment variables at all, but specifically I would like `:` to be allowed.

### PR DESCRIPTION
**Pull request by HirazawaUi**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
Allows users to set all ASCII characters with serial numbers in the range of 32-126 except "=" as environment variables.

Kubernetes should not restrict which environment variable names can be used, because it has no way of knowing what the application may need, and people can't always choose their own variable names, which may limit the adoption of Kubernetes.

#### Which issue(s) this PR fixes:

Fixes #53201

#### Does this PR introduce a user-facing change?

```release-note
Added a alpha feature, behind the `RelaxedEnvironmentVariableValidation` feature gate.
When that gate is enabled, Kubernetes allows almost all printable ASCII characters to be used in the names
of environment variables for containers in Pods.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/4369-allow-special-characters-environment-variable
```

### CODE REVIEW

File: pkg/apis/core/validation/validation.go [hunk 1]
```diff
@@ -2703,13 +2716,26 @@ func validateContainerResourceFieldSelector(fs *core.ResourceFieldSelector, expr
...
 		idxPath := fldPath.Index(i)
 		if len(ev.Prefix) > 0 {
-			for _, msg := range validation.IsEnvVarName(ev.Prefix) {
-				allErrs = append(allErrs, field.Invalid(idxPath.Child("prefix"), ev.Prefix, msg))
+			if utilfeature.DefaultFeatureGate.Enabled(features.RelaxedEnvironmentVariableValidation) {
+				for _, msg := range validation.IsRelaxedEnvVarName(ev.Prefix) {
+					allErrs = append(allErrs, field.Invalid(idxPath.Child("prefix"), ev.Prefix, msg))
+				}
+			} else {
+				for _, msg := range validation.IsEnvVarName(ev.Prefix) {
```
Comment (HirazawaUi) [inline pkg/apis/core/validation/validation.go:2729]: When the feature gate is disabled and the updated object cannot pass strict verification (relaxed verification will be used at this time), do we need to add a warning prompt?

**Comment by HirazawaUi** [comment]
@thockin @liggitt May you please review it? Since it's nearing the code freeze.

File: staging/src/k8s.io/apimachinery/pkg/util/validation/validation_test.go [hunk 2]
```diff
@@ -904,3 +904,28 @@ func TestIsDomainPrefixedPath(t *testing.T) {
...
+		"0a", "0 a", "'a", "(a", "@a",
+	}
+	for _, val := range goodValues {
+		if msgs := IsRelaxedEnvVarName(val); len(msgs) != 0 {
+			t.Errorf("expected true for '%s': %v", val, msgs)
+		}
+	}
+
+	badValues := []string{
+		"", "=", "a=", "1=a", "a=b", "#%=&&",
```
Comment (thockin) [inline staging/src/k8s.io/apimachinery/pkg/util/validation/validation_test.go:923]: put some non-printables ASCII and non-ASCII characters?

File: test/e2e/common/node/configmap.go [hunk 3]
```diff
@@ -258,6 +306,21 @@ func newConfigMap(f *framework.Framework, name string) *v1.ConfigMap {
 	}
 }
 
+func newSpecialKeyConfigMap(f *framework.Framework, name string) *v1.ConfigMap {
```
Comment (thockin) [inline test/e2e/common/node/configmap.go:309]: Can you comment this - what makes it special?  You could also just do this in-line - it's only used one time.

Please comment it

File: test/e2e/common/node/secrets.go [hunk 4]
```diff
@@ -238,6 +239,52 @@ var _ = SIGDescribe("Secrets", func() {
...
 		}
 	})
+
+	/*
+		Release: v1.30
+		Testname: Secrets, pod environment from source
+		Description: Create a secret. Create a Pod with Container that declares a environment variable using 'EnvFrom' which references the secret created to extract a key value from the secret.
+		Allow users to set Secret keys outside the C_IDENTIFIER scope as environment variables using EnvFrom.
+	*/
+	framework.It("should be consumed through special environment variable names", feature.RelaxedEnvironmentVariableValidation, func(ctx context.Context) {
```
Comment (thockin) [inline test/e2e/common/node/secrets.go:249]: "special" is not a helpful word

File: test/e2e/common/node/secrets.go [hunk 5]
```diff
@@ -254,6 +301,21 @@ func secretForTest(namespace, name string) *v1.Secret {
 	}
 }
 
+func specialKeySecretForTest(namespace, name string) *v1.Secret {
```
Comment (thockin) [inline test/e2e/common/node/secrets.go:304]: Same comments as above - just inline it.  And comment

File: staging/src/k8s.io/apimachinery/pkg/util/validation/validation.go [hunk 6]
```diff
@@ -431,6 +435,26 @@ func IsEnvVarName(value string) []string {
...
+
+	for i := range value {
+		asciiNum := value[i]
+		if !(unicode.IsPrint(rune(asciiNum)) && string(asciiNum) != "=" && asciiNum < unicode.MaxASCII) {
+			errs = append(errs, RegexError(RelaxedEnvVarNameFmtErrMsg, RelaxedEnvVarNameFmt, "my.env-name", "MY_ENV.NAME", "MyEnvName1"))
+			break
+		}
+	}
+
+	errs = append(errs, hasChDirPrefix(value)...)
```
[the remark this answers is not in the document]
Comment (HirazawaUi) [reply staging/src/k8s.io/apimachinery/pkg/util/validation/validation.go:454]: Yes, we don't need to check it.

File: pkg/apis/core/validation/validation.go [hunk 7]
```diff
@@ -2557,8 +2557,21 @@ func ValidateEnv(vars []core.EnvVar, fldPath *field.Path, opts PodValidationOpti
 		if len(ev.Name) == 0 {
 			allErrs = append(allErrs, field.Required(idxPath.Child("name"), ""))
 		} else {
-			for _, msg := range validation.IsEnvVarName(ev.Name) {
-				allErrs = append(allErrs, field.Invalid(idxPath.Child("name"), ev.Name, msg))
+			if utilfeature.DefaultFeatureGate.Enabled(features.RelaxedEnvironmentVariableValidation) {
```
[the remark this answers is not in the document]
Comment (HirazawaUi) [reply pkg/apis/core/validation/validation.go]: It has been fixed as per your suggestion, could you please review it again?

[the remark this answers is not in the document]
Comment (HirazawaUi) [reply pkg/apis/core/validation/validation.go]: ping @thockin

File: pkg/kubelet/kubelet_pods.go [hunk 8]
```diff
@@ -643,9 +643,17 @@ func (kl *Kubelet) makeEnvironmentVariables(pod *v1.Pod, container *v1.Container
 				if len(envFrom.Prefix) > 0 {
 					k = envFrom.Prefix + k
 				}
-				if errMsgs := utilvalidation.IsEnvVarName(k); len(errMsgs) != 0 {
-					invalidKeys = append(invalidKeys, k)
-					continue
+
+				if utilfeature.DefaultFeatureGate.Enabled(features.RelaxedEnvironmentVariableValidation) {
```
Comment (HirazawaUi) [inline pkg/kubelet/kubelet_pods.go:647]: This seems to bring a risk, if the user closes the feature gate and has used relaxed validation before update (it can still pass top-level validation), but the start container will fail here.

Comment (thockin) [reply pkg/kubelet/kubelet_pods.go:647]: I think the feature gate really just controls admission.  There are a few archetype changes that need gates:

For changes which add a net-new API field, that means the REST code (pkg/registry/...) does:
```
if gate_disabled && !already_in_use {
    wipe the field
}
```
Validation and impl code don't really need to check the gate.

For changes which don't add a new field, but change what's allowed in the API (e.g. new enums, loosening validation), the validation code does:
```
if gate_disabled && !already_in_use {
    validation error
}
```
Impl code doesn't really need to check the gate.

For changes which don't change the API, but change how things operate, the impl code does:
```
if gate_enabled {
    do something new
} else {
    do something old
}
```

@liggitt I have been meaning to write this down ....somewhere.  Do you agree with the general positioning?  Currently it's done however people think makes sense, with no consistency.

Comment (thockin) [reply pkg/kubelet/kubelet_pods.go:647]: For this specific case, I am unsure.  Kubelet should not really need to re-validate pods it gets from the apiserver, but I guess it is possible for the apiserver to feed a value that kubelet does not understand.  More important is static pods - someone has to validate those.  Unfortunately the validation code is all internal-version, so not exactly reusable unless we want kubelet to use internal-versions.

I am not as familiar with this path.

@tallclair or @SergeyKanzhelev - what's the norm here?  How is it supposed to work?    

If we do not validation here, what is the failure mode - static pod sets `name: foo=bar` and runtime fails to create the pod?

File: pkg/kubelet/kubelet_pods.go [hunk 9]
```diff
@@ -678,9 +686,17 @@ func (kl *Kubelet) makeEnvironmentVariables(pod *v1.Pod, container *v1.Container
...
```
Comment (thockin) [inline pkg/kubelet/kubelet_pods.go:690]: same here

File: test/e2e/feature/feature.go [hunk 10]
```diff
@@ -253,6 +253,9 @@ var (
 	// TODO: document the feature (owning SIG, when to use this feature for a test)
 	RecoverVolumeExpansionFailure = framework.WithFeature(framework.ValidFeatures.Add("RecoverVolumeExpansionFailure"))
 
+	// TODO: document the feature (owning SIG, when to use this feature for a test)
```
[the remark this answers is not in the document]
Comment (HirazawaUi) [reply test/e2e/feature/feature.go:256]: now :)

File: pkg/kubelet/kubelet_pods.go (hunk 8 above)
Comment (liggitt) [reply pkg/kubelet/kubelet_pods.go:647]: > For changes which add a net-new API field, that means the REST code (pkg/registry/...) does:
> 
> ```
> if gate_disabled && !already_in_use {
>     wipe the field
> }
> ```
> 
> Validation and impl code don't really need to check the gate.

I agree with that. That's roughly what is described in https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#new-field-in-existing-api-version but it's *very* well buried.

> For changes which don't add a new field, but change what's allowed in the API (e.g. new enums, loosening validation), the validation code does:
> 
> ```
> if gate_disabled && !already_in_use {
>     validation error
> }
> ```

Rather than validation code checking the diff and feature gate directly, I would model this as validation options passed into the Validate() and ValidateUpdate() functions. That makes it easier to unit test and easier to centralize the decision about when to allow the new value in the strategy. That is also documented in https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#new-enum-value-in-existing-field but could certainly improve its visibility and clarity.



> For changes which don't change the API, but change how things operate, the impl code does:
> 
> ```
> if gate_enabled {
>     do something new
> } else {
>     do something old
> }
> ```

Agreed.

Comment (liggitt) [reply pkg/kubelet/kubelet_pods.go:647]: > More important is static pods - someone has to validate those. Unfortunately the validation code is all internal-version, so not exactly reusable unless we want kubelet to use internal-versions.

That's exactly what kubelet does:

https://github.com/kubernetes/kubernetes/blob/8674282a054d3ae32e2e009dab6f8a0da3689828/pkg/kubelet/config/common.go#L109-L191

Comment (HirazawaUi) [reply pkg/kubelet/kubelet_pods.go:647]: > If we do not validation here, what is the failure mode

CRI accepts a wider range of characters as env names than k8s, and if we just don't validate the envvar name here, it shouldn't fail here

Since we have validated the static pod using the [validation code](https://github.com/kubernetes/kubernetes/blob/19df52c41a25cedb977b1dc064cdaa916261de67/pkg/kubelet/config/common.go#L110-L142), I think we can use the `IsRelaxedEnvVarName` function here to validate it?

I will seek the help of the SIG Node maintainers at slack, their advice would definitely be better...

File: pkg/apis/batch/validation/validation.go [hunk 11]
```diff
@@ -139,11 +139,16 @@ func validateGeneratedSelector(obj *batch.Job, validateBatchLabels bool) field.E
 }
 
 // ValidateJob validates a Job and returns an ErrorList with any errors.
-func ValidateJob(job *batch.Job, opts JobValidationOptions) field.ErrorList {
+func ValidateJob(job, oldJob *batch.Job, opts JobValidationOptions) field.ErrorList {
```
Comment (kannon92) [inline pkg/apis/batch/validation/validation.go:142]: I don't really understand why this change is needed for this PR?

File: test/e2e/common/node/configmap.go [hunk 12]
```diff
@@ -242,6 +243,54 @@ var _ = SIGDescribe("ConfigMap", func() {
 		framework.ExpectNoError(err, "failed to list ConfigMap by LabelSelector")
 		gomega.Expect(configMapList.Items).To(gomega.BeEmpty(), "ConfigMap is still present after being deleted by collection")
 	})
+
+	/*
+		Release: v1.30
```
Comment (kannon92) [inline test/e2e/common/node/configmap.go:248]: How are you testing this feature considering that the gate is off by default?

This shouldn't block the PR imo as we have test freeze for this.

**Comment by kannon92** [comment]
please update the release note.

/release-note-edit added an alpha feature, behind the `RelaxedEnvironmentVariableValidation` feature gate.
When that gate is enabled, Kubernetes allows almost all printable ASCII characters to be used in the names
of environment variables for containers in Pods.

File: pkg/kubelet/kubelet_pods.go (hunk 8 above)
Comment (kannon92) [reply pkg/kubelet/kubelet_pods.go:647]: So static pods is one case and I've seen mentions in the code of standalone Kubelet. I don't really know how much that is used (maybe just for static pods). So I think validation in kubelet makes sense.

I think reusing the function makes sense to me.

File: test/e2e/common/node/configmap.go (hunk 12 above)
Comment (HirazawaUi) [reply test/e2e/common/node/configmap.go:248]: We use feature.RelaxedEnvironmentVariableValidation as a parameter. The current test will only be run when the RelaxedEnvironmentVariableValidation feature gate is turned on.

File: pkg/apis/batch/validation/validation.go (hunk 11 above)
Comment (HirazawaUi) [reply pkg/apis/batch/validation/validation.go:142]: We need all old resource objects containing PodSpec so that we can judge whether the updating object use relaxed or strict validation mode.

File: test/e2e/common/node/configmap.go (hunk 12 above)
Comment (kannon92) [reply test/e2e/common/node/configmap.go:248]: feature.RelaxedEnvironmentVariableValidation is not related to the feature gate.

Its a name given to a test so you can run the test using a Focus.

File: pkg/apis/batch/validation/validation.go (hunk 11 above)
Comment (kannon92) [reply pkg/apis/batch/validation/validation.go:142]: Sounds good.

File: pkg/kubelet/kubelet_pods.go (hunk 8 above)
Comment (HirazawaUi) [reply pkg/kubelet/kubelet_pods.go:647]: > I think are suggesting setting the feature gate as an option for validation and removing the feature gates from validation logic.

I don't quite understand here, may you elaborate?

Comment (kannon92) [reply pkg/kubelet/kubelet_pods.go:647]: You should not have feature gates in any validation code and just instead set the PodValidationOptions.

See:

```
// PodValidationOptions contains the different settings for pod validation
type PodValidationOptions struct {
	// Allow invalid pod-deletion-cost annotation value for backward compatibility.
	AllowInvalidPodDeletionCost bool
	// Allow invalid label-value in LabelSelector
	AllowInvalidLabelValueInSelector bool
	// Allow pod spec to use non-integer multiple of huge page unit size
	AllowIndivisibleHugePagesValues bool
	// Allow pod spec to use status.hostIPs in downward API if feature is enabled
	AllowHostIPsField bool
	// Allow invalid topologySpreadConstraint labelSelector for backward compatibility
	AllowInvalidTopologySpreadConstraintLabelSelector bool
	// Allow node selector additions for gated pods.
	AllowMutableNodeSelectorAndNodeAffinity bool
	// Allow projected token volumes with non-local paths
	AllowNonLocalProjectedTokenPath bool
	// Allow namespaced sysctls in hostNet and hostIPC pods
	AllowNamespacedSysctlsForHostNetAndHostIPC bool
	// The top-level resource being validated is a Pod, not just a PodSpec
	// embedded in some other resource.
	ResourceIsPod bool
}
```

You can use the feature gate to see an option and just reference that option. This makes it easier for testing and you can allow for skew a bit easier.

Comment (thockin) [reply pkg/kubelet/kubelet_pods.go:647]: > Rather than validation code checking the diff and feature gate directly, I would model this as validation options passed into the Validate() and ValidateUpdate() functions. 

OK.  That explains the pattern which I have seen but didn't really look into why we were doing it that way.

File: pkg/apis/core/validation/validation.go [hunk 13]
```diff
@@ -2549,23 +2549,52 @@ func validateContainerPorts(ports []core.ContainerPort, fldPath *field.Path) fie
...
 	allErrs := field.ErrorList{}
 
 	for i, ev := range vars {
 		idxPath := fldPath.Index(i)
 		if len(ev.Name) == 0 {
 			allErrs = append(allErrs, field.Required(idxPath.Child("name"), ""))
 		} else {
-			for _, msg := range validation.IsEnvVarName(ev.Name) {
-				allErrs = append(allErrs, field.Invalid(idxPath.Child("name"), ev.Name, msg))
+			if utilfeature.DefaultFeatureGate.Enabled(features.RelaxedEnvironmentVariableValidation) || relaxedEnvVarUsed(ev.Name, oldPodEnvVarNames) {
```
Comment (thockin) [inline pkg/apis/core/validation/validation.go:2560]: For completeness, check the "options" here rater than the gate.  I think this will remove the need to do a lot of the plumbing you did (I'm so sorry - I didn't know this pattern).

Specifically, look at pkg/registry/core/pod/strategy.go `podStrategy.Validate)`.  It does:

```
    opts := podutil.GetValidationOptionsFromPodSpecAndMeta(&pod.Spec, nil, &pod.ObjectMeta, nil)
    opts.ResourceIsPod = true
    return corevalidation.ValidatePodCreate(pod, opts)
```

For things like Job, we already do:
```
opts := validationOptionsForJob(job, oldJob)
```

where `JobValidationOptions` embeds `PodValidationOptions`

So aaaaaaallllll those places you fixed to pass "old" maybe don't need it.  Instead you check this in pkg/api/pod/util.go

Comment (HirazawaUi) [reply pkg/apis/core/validation/validation.go:2560]: Okay, I'll finish it soon.

**Comment by HirazawaUi** [comment]
@thockin @kannon92  Changes have been completed as suggested.

File: staging/src/k8s.io/apimachinery/pkg/util/validation/validation.go [hunk 14]
```diff
@@ -418,6 +419,9 @@ func IsHTTPHeaderName(value string) []string {
 const envVarNameFmt = "[-._a-zA-Z][-._a-zA-Z0-9]*"
 const envVarNameFmtErrMsg string = "a valid environment variable name must consist of alphabetic characters, digits, '_', '-', or '.', and must not start with a digit"
 
+// TODO(hirazawaui): Rename this when the RelaxedEnvironmentVariableValidation gate is removed.
+const relaxedEnvVarNameFmtErrMsg string = "a valid environment variable names must be printable ASCII characters other than '=' character"
```
Comment (thockin) [inline staging/src/k8s.io/apimachinery/pkg/util/validation/validation.go:423]: Grammar - this could be a followup.

 `a valid environment variable name must consist only of printable ASCII characters other than '='`

File: staging/src/k8s.io/apimachinery/pkg/util/validation/validation.go [hunk 15]
```diff
@@ -431,6 +435,24 @@ func IsEnvVarName(value string) []string {
 	return errs
 }
 
+// IsRelaxedEnvVarName tests if a string is a valid environment variable name.
+func IsRelaxedEnvVarName(value string) []string {
+	var errs []string
+
+	if len(value) == 0 {
+		errs = append(errs, "environment variable name"+EmptyError())
```
Comment (thockin) [inline staging/src/k8s.io/apimachinery/pkg/util/validation/validation.go:443]: need a trailing space before EmptyError:

Comment (thockin) [reply staging/src/k8s.io/apimachinery/pkg/util/validation/validation.go:443]:
```suggestion
		errs = append(errs, "environment variable name "+EmptyError())
```

**Comment by thockin** [comment]
I'm going to approve - my comments are OK for a followup.

**Comment by thockin** [comment]
/approve
/lgtm

FOLLOWUP REQUIRED

File: pkg/kubelet/kubelet_pods.go (hunk 8 above)
Comment (HirazawaUi) [reply pkg/kubelet/kubelet_pods.go]: @thockin 
If we don't verify here, when we use `EnvFrom` in the pod spec, we will allow `configmap/secret` keys that were not allowed before, because `EnvFrom` allows us to directly reference the keys in `configmap/secret` as environment variable names, `configmap/secrets` keys character range  is wider than that of envvar names, the difference is that envvar is not allowed to start with a digit.

Previously, we would verify whether the keys of `configmap/secret` are legal envvar names. If not, we would skip setting this environment variable for the container, output events, and start the container.
ref: https://kubernetes.io/docs/concepts/configuration/secret/#restriction-env-from-invalid

Now we allow all `configmap/secret` keys as envvar names.

According to the [comments](https://github.com/kubernetes/kubernetes/pull/123385#discussion_r1509694339), we shouldn't check the feature gate here (maybe we should), so I modified some documentation for this question, may you take a look.
